### PR TITLE
Add kernel momentum 3-block NMD from Seraghiti et.al. (2023) 

### LINKED ISSUE #13: [Feature Request] Add kernel momentum 3-block NMD from Seraghiti et.al. (2023)
**Which of these best describes your feature request:**
- [ ] Library usability improvement
- [ ] Performance improvement (speed, accuracy)
- [x] New methodology or algorithm

**Describe how the new feature would improve the library:**
I suggest the addition of the momentum 3-block NMD kernel, as described in [Seraghiti et.al. (2023)](https://arxiv.org/abs/2305.08687). It is an expansion of the current base model-free kernel as it also alternates between constructing utility matrix Z and the low-rank approximation. In contrast to the base model-free kernel: 

- it **finds the low-rank approximation's factors W and H sequentially** by solving matrix least-squares problems instead of using tSVD for direct recovery. 
- To accelerate convergence, they **add momentum steps where they extrapolate matrix Z and the low-rank candidate L** using a Polyak-type extrapolation with fixed momentum parameter $\beta$.

**Describe the solution you'd like**
Leaving aside most convergence criteria, their matlab implementation [2] uses the following input parameters:

- X: sparse non-negative matrix of shape (m, n)
- r: target rank of the low-rank approximation
- W0: candidate low-rank factor (m, r)
- H0: candidate low-rank factor (r, n)
- $\beta$: fixed momentum parameter
- maxiters: maximum number of iterations, to avoid momentum step on L (called $\Theta$ in [1]) in the last iteration as it changes matrix rank.

And returns the low-rank factors W, H.

Following challenges come with their implementation:
- Kernel initialized by W0 & H0, instead of L: we might need to alter standard inputs in `KernelInputType`
- Kernels in nomad don't have access to a parameter `maxiters`: I think we can omit maxiters with only knowing the current iteration number by performing the momentum step on L at the beginning of the algorithm starting in the second iteration.

My idea for the step method:
```{python}
def step(self) -> None:
	if self.elapsed_iterations > 0:
		self.low_rank_candidate_L = apply_momentum(
			self.low_rank_candidate_L, 
			self.previous_low_rank_candidate_L, 
			self.momentum_beta
		)

	utility_matrix_Z = construct_utility(
	  self.low_rank_candidate_L, 
	  self.sparse_matrix_X
	)

	self.utility_matrix_Z = apply_momentum( 
		utility_matrix_Z, 
		self.utility_matrix_Z, 
		self.momentum_beta
	)

	self.previous_low_rank_candidate_L = self.factor_W @ self.factor_H

	self.factor_W = update_W(self.factor_H, self.utility_matrix_Z)
	self.factor_H = update_H(self.factor_W, self.utility_matrix_Z)

	self.low_rank_candidate_L = self.factor_W @ self.factor_H
```
**Utility functions**
Function `construct_utility` is shared with the base model-free kernel, `apply_momentum` can be reused in the Accelerated Momentum NMD algorithm [1]. I'd rename `base_model_free_util.py` into `model_free_util.py` for functions that are shared across the model-free kernels and add, e.g. `update_W` to `3b_model_free_util.py`. 

**kernelInputTypes**
- Additional hyperparameter $\beta$ 
- Instantiated with W0, H0 instead of a low-candidate L

I'm unfamiliar with the way the input types are defined, specifically kernel-specific parameters. For the sake of simplicity, let's pretend 3B-NMD would be instantiated with L as well, would the following work?

```{python}
class Momentum3BModelFreeInputType(KernelInputType):
    beta: float

KernelSpecificParameters = Union[float, int, Momentum3BModelFreeInputType]
```

**kernelReturnTypes**
3B-NMD returns W & H, so adds nothing to the base return type

**References**
- \[1\] Seraghiti, G., et. al. (2023), "Accelerated Algorithms for Nonlinear Matrix Decomposition with the ReLU Function" [https://doi.org/10.48550/arXiv.2305.08687](https://doi.org/10.48550/arXiv.2305.08687)|
- \[2\] [Matlab Implementation](https://gitlab.com/ngillis/ReLU-NMD/)

### PR DESCRIPTION
**Pull request by sfohr**
Add Momentum 3-Block model-free kernel

Closes #13 .

### Type of change
- [ ] Bug fix (non-breaking change which fixes an unintended
or erroneous behavior)
- [x] New features (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that changes how existing
functionality works, other than correcting errors)
- [ ] Code improvement (no new functionality, but improving project structure)
- [ ] Documentation only (changes that improve or expand instructions
to users, without changing program behavior)

## Motivation and Context
The change adds the Momentum 3-Block model-free kernel described in [Seraghiti et. al. 2023](https://arxiv.org/abs/2305.08687) ([their matlab implementation](https://gitlab.com/ngillis/ReLU-NMD/)). It's an extension of the base model-free algorithm from Saul (2022) and alters/adds the base algorithm in two ways:

1. It uses the parametrization $L=WH$ to update $W$ and $H$ separately by solving least-squares problems and thus avoids computation of the rank-r tSVD, which reduces computational complexity from $O(mnr^2)$ to $O(mnr)$
2. Extrapolation of utility $Z$ and low-rank candidate $L$ using a momentum term with fixed parameter `momentum_beta` to accelerate convergence.

## Description

This PR introduces the following changes:

* Moved functions `construct_utility` and `apply_momentum` from file `base_model_free_util.py` to `model_free_util.py`
  * Functions will be used by multiple model-free kernels (base, momentum 3-block and aggressive momentum NMD)

* Added kernel-specific utility functions `update_W`, `update_H` and `_compute_least_squares_solution` replacing direct factorization of `Z` via truncated SVD
  * `_compute_least_squares_solution`: given two matrices `A` and `B` solves $argmin_{X} || B - A @ X ||^{2}_{F}$, used internally in `update_W` and `update_H`
  * `update_W`: updates candidate factor `W`, given `Z` and `H` by solving: $argmin_{W} || Z - W @ H ||^{2}_{F}$.
  * `update_H`: updates candidate factor `H`, given `Z` and `W` by solving: $argmin_{H} || Z - W @ H ||^{2}_{F}$.
  * affected file `momentum_three_block_model_free_util.py`
* Added kernel-specific arguments in `Momentum3BlockAdditionalParameters`
  * `momentum_beta`: determines the magnitude of the momentum terms on `Z`and `L`, defaults to 0.7
  *  `candidate_factor_W0`: initial candidate for W, optional.
  * `candidate_factor_H0`: initial candidate for H, optional.
  * The candidate factors are optional and only used if BOTH are provided. If at least one of them is missing, initial candidates for `W` and `H` are generated by factorizing `low_rank_candidate_L` with a rank-r truncated SVD. So the user can utilize initialization options that return a candidate for `L`.
  * affected file: `kernelInputTypes.py`
* Added function `two_part_factor_known_rank` to `decomposition_util`
  * Used when the kernel is instantiated without both `W` and `H` - factorizes `L` (n, p) into candidates `W`(n, r) and `H`(r, p)
  * This is a similar to `two_part_factor` but rank is supplied directly.
  * We could actually get rid of `two_part_factor` and only use `two_part_factor_known_rank` as we always know the desired rank.
  * affected file: `decomposition_util.py`
* Added a kernel-specific return type `Momentum3BlockModelFreeKernelReturnType`
  * Subclass of `KernelReturnBase`, adds nothing to base.
  * affected file: `kernelReturnTypes.py`
* Added `MOMENTUM_3_BLOCK_MODEL_FREE` kernel strategy to `KernelStrategy`
  * affected file: `enums.py`
* Implemented the kernel strategy in `instantiate_kernel`
  * Removed `NotImplementedError` if kernels-specific parameters are given
  * Added a type check on the kernel-specific parameter object before instantiating the kernel to calm pylint (+removed test `test_instantiate_kernel_throws_on_unsupported_feature`)
  * affected files: `factory_util.py`, `test_factory_util.py`

### Utility Functions
Tests the computations performed in the kernel.

* `_compute_least_squares_solution`
  * We reverse the computation
  * Then function solves `argmin_{X} || B - A @ X ||^{2}_{F}`
  * Given matrices `A` and `X`, we compute $B=A*X$
  * Compute our actual result with `_compute_least_squares_solution(A, B)`
  * And assert that `actual_result` is almost equal to `X`(our expected result)
  * Test: `test_compute_least_squares_solution` in  file `test_momentum_three_block_model_free_util.py`

* `update_W`
  * Use alternative method to solve the problem
  * Calculate expected result using the inverse: `(np.linalg.inv(H @ H.T) @ H @ Z.T).T`
  * Compute actual result `update_W(H, Z)` and assert that actual is almost equal to expected result
  * a better solution might be to mock `_compute_least_squares_solution` and check if the function was called with arguments `H@H.T` and `H@Z.T`, but I could not figure out how to achieve this, as pytest seems to test instance equality instead of checking the individual array values for equality.

* `update_H`
  * Use alternative method to solve the problem
  * Calculate expected result using the inverse: `np.linalg.inv(W.T @ W) @ W.T @ Z`
  * Compute actual result `update_H(W, Z)` and assert that actual is almost equal to expected result

### Kernel
Mostly tests the logic of the kernel - if utility functions are called the desired amount of times with the desired arguments.

* `test_momentum_3block_model_free_kernel_step_W0H0_given_one_iter`
  * test kernel logic for one iteration (so momentum step on L is not performed)
* `test_momentum_3block_model_free_kernel_step_apply_momentum_call_count`
  * test kernel logic for two iterations
  * assure that `apply_momentum` is called exactly 3 times:
    * once in first iteration
    * twice in second iteration
* `test_momentum_3block_model_free_kernel_W0H0_initialized_if_not_given`
  * test if W0, H0 are computed as rank-r truncated SVD of `low_rank_candidate_L`
  * initialize kernel without W0, H0, kernel is supposed to compute `W0`and `H0`from `L`
  * then assert that product of kernel variables is (almost) equal to `low_rank_candidate_L`
* `test_momentum_3block_model_free_kernel_running_report`
  * same test as in the base model-free case
* `test_momentum_3block_model_free_kernel_final_report`
  * same test as in the base model-free case
* `test_momentum_3block_model_free_kernel_per_iteration_diagnostic_calls_base`
  * same test as in the base model-free case

### Integration Test
Replaced the `all_kernels` object to `all_kernels_with_params` to accomodate kernel-specific parameters if necessary. Adapted the existing integration tests accordingly.

## Checklist

- [x] My code follows the project code style
  - [x] I've run the automatic formatters and made any needed changes
  - [x] I've ensured that my new code is explained in docstrings/comments
- [ ] My submission requires a change to the documentation
  - [ ] I've updated the documentation to reflect the changes
- [x] I've added unit tests that cover the visible behavior of the changes
- [x] I've added or modified integration tests that ensure the whole system
works on a practical problem after my changes

## Closing thoughts

As discussed, the branch includes changes suggested in #14 .

Kernel methods `running_report` and `report` are shared between the base model-free, momentum 3-block and aggressive momentum kernels. Might make sense to move them to `KernelBase`.

### CODE REVIEW

File: src/fi_nomad/util/decomposition_util.py [hunk 1]
```diff
@@ -138,3 +140,23 @@ def two_part_factor(matrix: FloatArrayType) -> Tuple[FloatArrayType, FloatArrayT
     part_A = svd_U @ inner_a
     part_B = np.pad(np.eye(rank), ((0, 0), (0, matrix.shape[1] - rank))) @ svd_Vt
     return (part_A, part_B)
+
+
+def two_part_factor_known_rank(
+    matrix: FloatArrayType, rank: int
+) -> Tuple[FloatArrayType, FloatArrayType]:
+    """Factor matrix into two rectangular matrices with inner dimension `rank`
```
Comment (jsoules) [inline src/fi_nomad/util/decomposition_util.py:148]: I like your suggestion to merge this with the original `two_part_factor`. I think the easiest thing to do would be to make the `rank` parameter `Optional[int]` and then just compute it using `np.linalg.matrix_rank` if it is `None`/not defined.

I expect you're probably right that we can replace all calls to this function with ones that explicitly state the target rank, but let's do that in a separate issue.

File: src/fi_nomad/util/model_free_util.py [hunk 2]
```diff
@@ -3,8 +3,10 @@
```
Comment (jsoules) [inline src/fi_nomad/util/model_free_util.py:3]: The docstring probably needs to be changed, since we are now expanding the functions defined in this file :smile:

File: src/fi_nomad/util/model_free_util.py [hunk 3]
```diff
@@ -38,3 +40,36 @@ def construct_utility(
...
+        Updated matrix X after applying the momentum term.
+
+    Example:
+        # Setting beta to 1 will double the step taken from the last iteration to
+        # the current.
+        X = apply_momentum(X, X_previous, beta=1.0)
+
+        # Setting beta to 0 will disable the momentum step and just return X
+        post_momentum_X = apply_momentum(X, X_previous, beta=1.0)
+        assert np.array_equal(X, post_momentum_X)
```
Comment (jsoules) [inline src/fi_nomad/util/model_free_util.py:73]: Shouldn't this be `beta=0.0`?

File: test/test_util/test_factory_util.py [hunk 4]
```diff
@@ -30,17 +36,6 @@ def test_insantiate_kernel_throws_on_instantiation_failure(mock_base: Mock) -> N
...
-        _ = instantiate_kernel(
-            KernelStrategy.BASE_MODEL_FREE,
-            mock_data_in,
-            kernel_params=some_non_null_value,
-        )
-
-
 @patch(f"{PKG}.RowwiseVarianceGaussianModelKernel", autospec=True)
 @patch(f"{PKG}.SingleVarianceGaussianModelKernel", autospec=True)
 @patch(f"{PKG}.BaseModelFree", autospec=True)
```
Comment (jsoules) [inline test/test_util/test_factory_util.py:41]: I'd suggest adding a couple lines to `test_instantiate_kernel_honors_strategy_selection` to test instantiating the kernel with `KernelStrategy.MOMENTUM_3_BLOCK_MODEL_FREE`--it's a bit silly, but it ensures complete test coverage without running the integration tests.

File: test/test_util/test_model_free_util.py [hunk 5]
```diff
@@ -29,3 +27,24 @@ def test_construct_utility() -> None:
...
+    ])
+    momentum_parameter = 0.5
+
+    expected_matrix = np.array([
+        [2, 6, 1],
+        [0, 3, -0.5]
+    ])
+    # fmt: on
+    result = apply_momentum(current_matrix, previous_matrix, momentum_parameter)
+    np.testing.assert_array_equal(expected_matrix, result)
```
Comment (jsoules) [inline test/test_util/test_model_free_util.py:50]: The comments for the `apply_momentum` function suggest two more invariants:
```python
beta = 1.0
result = apply_momentum(current_matrix, previous_matrix, beta)
np.testing.assert_array_equal(result - current_matrix, 2 * (current_matrix - previous_matrix))
```
and
```python
beta = 0.0
result = apply_momentum(current_matrix, previous_matrix, beta)
assert_array_equal(result, current)
```

I am confident that the function works as intended, but it's nice to check the stated invariants just in case something weird happens.

File: test/test_util/test_momentum_three_block_model_free_util.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    Z = np.array([
+        [5.0, 6.0], 
+        [7.0, 8.0]
+    ])
+
+    # fmt: off
+    expected_result = np.linalg.inv(H @ H.T) @ H @ Z.T
+    expected_result = expected_result.T
+    actual_result = update_W(H, Z)
+    np.testing.assert_array_almost_equal(actual_result, expected_result)
```
Comment (jsoules) [inline test/test_util/test_momentum_three_block_model_free_util.py:45]: You mentioned the possibility of mocking the calls to `_compute_least_squares_solution`. In principle, sure; but in practice, this is a sufficiently well-defined math operation that I'm not too worried about it. That said, if you'd like to write it with a mock and are having some trouble finding the right way, I'm happy to help out--just let me know.

File: test/test_kernels/test_momentum_three_block_model_free_kernel.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+def test_momentum_3block_model_free_kernel_W0H0_initialized_if_not_given(
+    fixture_no_tol_W0H0_not_given: Fixture,
+) -> None:
+    (indata, _, kernel) = fixture_no_tol_W0H0_not_given
+
+    np.testing.assert_array_almost_equal(
+        indata.low_rank_candidate_L,
+        kernel.candidate_factor_W @ kernel.candidate_factor_H,
+    )
```
Comment (jsoules) [inline test/test_kernels/test_momentum_three_block_model_free_kernel.py:135]: How annoying would it be to test this for the cases where only one of `W0` and `H0` is defined? Again I'm not super worried about it, and if it requires setting up two more whole fixtures or radically reworking how these are structured, it probably isn't worth it; just noting that that is an edge case which isn't covered.

File: src/fi_nomad/kernels/momentum_three_block_model_free.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+        Returns:
+            Object containing results and summary
+        """
+        floss = str(self.loss) if self.loss != float("inf") else "Not Tracked"
+        text = f"{self.elapsed_iterations} total, final loss {floss}"
+        data = Momentum3BlockModelFreeKernelReturnType(
+            (self.candidate_factor_W, self.candidate_factor_H)
+        )
+        return KernelReturnType(text, data)
```
Comment (jsoules) [inline src/fi_nomad/kernels/momentum_three_block_model_free.py:144]: I saw your note that this is repeated and could be moved to the base class. That's a good idea. The only tricky part will be adding the `data` part, since it's not easy to do one-line differences in the middle of functions. But we can either make it work with inheritance or restructure how the report call gets made.

Let's make that cleanup a separate issue--I don't want this PR to be held up just for solving that.

**Review by jsoules — COMMENTED** [review]
This is good work and just about ready. I have a few comments suggesting some minor revisions--please let me know if you have any questions, or would like any support.

Otherwise this looks good to merge, assuming you are satisfied with the performance of the algorithm as implemented.

**Comment by sfohr** [comment]
> This is good work and just about ready. I have a few comments suggesting some minor revisions--please let me know if you have any questions, or would like any support.
> 
> Otherwise this looks good to merge, assuming you are satisfied with the performance of the algorithm as implemented.

Thank you for the review. We can merge - I'll adress the suggested changes in a separate PR.

**Comment by jsoules** [comment]
Ok, happy to merge as-is and fix the issues in a later PR.

However, we do need to make one small tweak. There's a conflict with the base branch `trunk`--it looks like this is just git being silly about the imports in `src/fi_nomad/types/kernelInputTypes.py`, since your branch imports `Optional` but the updated `trunk` does not.

Can you pull the current `trunk` and then either merge it or rebase off it? (Rebase leads to a cleaner commit history and is perfectly safe if you're careful, but you would have to do a force push to update your current branch--so I leave it up to you.)

Also, the `pylint` check is failing because of the tiny bit of repeated code (that we've already noted). I updated the configuration in `trunk` so that `pylint` will still return a passing exit code with a score above 9.5, so that issue should go away.
